base/intfuncs.jl

```diff
@@ -744,17 +744,25 @@ julia> digits!([2,2,2,2,2,2], 10, base = 2)
 ```
 """
 function digits!(a::AbstractVector{T}, n::Integer; base::Integer = 10) where T<:Integer
-    base < 0 && isa(n, Unsigned) && return digits!(a, convert(Signed, n), base = base)
     2 <= abs(base) || throw(ArgumentError("base must be ≥ 2 or ≤ -2, got $base"))
     hastypemax(T) && abs(base) - 1 > typemax(T) &&
         throw(ArgumentError("type $T too small for base $base"))
-    for i in eachindex(a)
-        if base > 0
-            a[i] = rem(n, base)
-            n = div(n, base)
-        else
-            a[i] = mod(n, -base)
-            n = cld(n, base)
+    isempty(a) && return a
+
+    if base > 0
+        for i in eachindex(a)
+            n, d = divrem(n, base)
+            a[i] = d
+        end
+    else
+        # manually peel one loop iteration for type stability
+        n, d = fldmod(n, -base)
+        a[firstindex(a)] = d
+        n = -signed(n)
```

Comment:
This is type unstable, but maybe the compiler is good enough to handle that, or that the the cost of the divisions makes this negligible?

Reply:
This used to be type unstable but since the introduction of SSAIR the compiler doesn't care about the reuse of variable names at all (that's basically what the SS part means: single static assignment). It only matters if the same "slot" in a basic block can have different types, e.g. in the loop body below. Even then the union splitting can deal with it fairly often.

Reply:
Thanks for this explanation, that's very good to know, it allows not bothering with this kind of gymnastic!